File: src/main/java/com/kata/bank_account/domain/entities/BankAccount.java
package com.kata.bank_account.domain.entities;

import com.kata.bank_account.domain.exception.InsufficientFundsException;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public abstract class BankAccount {
    protected final UUID id;
    protected final List<BankTransaction> transactions = new ArrayList<>();

    protected BankAccount() {
        id = UUID.randomUUID();
    }

    public void deposit(BigDecimal amount) {
        this.transactions.add(new BankTransaction(amount));
    }

    public void withdraw(BigDecimal amount){
        if (canWithdraw(amount)) {
            this.transactions.add(new BankTransaction(amount.negate()));
            return;
        }
        throw new InsufficientFundsException();
    }

    protected boolean canWithdraw(BigDecimal amount) {
        return balance().compareTo(amount) >= 0;
    }

    public BigDecimal balance() {
        return transactions
                .stream()
                .map(BankTransaction::amount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    protected abstract boolean canOverdraft();
}

File: src/main/kotlin/com/kata/bank/BankAccount.kt
package com.kata.bank

import java.util.*

class BankAccount {
    lateinit var id: UUID
    private var transactions: Transactions = Transactions()

    fun balance(): Long {
        return transactions.total()
    }

    fun deposit(amount: Long) {
        transactions.add(Transaction(amount))
    }

    fun withdraw(amount: Long) {
        if (amount > balance()) {
            throw KInsufficientFundsException()
        }
        transactions.add(Transaction(-amount))
    }
}